storage/client_range_test.go

```diff
@@ -150,3 +155,180 @@ func TestRejectFutureCommand(t *testing.T) {
 		t.Errorf("expected 15, got %v", v)
 	}
 }
+
+// TestTxnPutOutOfOrder tests a case where a put operation of an older
+// timestamp comes after a put operation of a newer timestamp in a
+// txn. The test makes sure such an out-of-order put succeeds and
+// overrides an old value.
+//
+// The test uses a "Writer" and a "Reader" to reproduce an
+// out-of-order put. The Writer executes a put operation in a txn and
+// commits it. The Reader executes two get operations on the same key
+// with higher priority (outside of a txn), which pushes the Writer's
+// intent forward in time, so that the Writer's retry happens in the
+// previous attempt's past.
+//
+// 1) The Writer executes a put operation with time T in a txn.
+// 2) Before the Writer's txn is committed, the Reader sends a get
+//    operation with time T+100. This pushes the Writer txn timestamp to
+//    T+100 and triggers the restart of the Writer's txn. A write intent
+//    at timestamp T+100 is written.
+// 3) Before the Writer starts a new epoch of the txn, the Reader sends
+//    another get operation with time T+200. The Writer txn timestamp is
+//    pushed again to T+200 by updating the timestamp of the write
+//    intent to T+200. We fail the actual get operation, so that
```

Comment:
The test deliberately fails the get operation, so cockroach doesn't update its read timestamp cache.